pkg/osquery/tables/systemprofiler/systemprofiler.go

```diff
@@ -0,0 +1,236 @@
+
+var knownDetailLevels = []string{
+	"mini",  // short report (contains no identifying or personal information)
+	"basic", // basic hardware and network information
+	"full",  // all available information
+}
+
+type Property struct {
+	Order                string `plist:"_order"`
+	SuppressLocalization string `plist:"_suppressLocalization"`
+	DetailLevel          string `plist:"_detailLevel"`
+}
+
+type Result struct {
+	Items          []interface{} `plist:"_items"`
+	DataType       string        `plist:"_dataType"`
+	SPCommandLine  []string      `plist:"_SPCommandLineArguments"`
+	ParentDataType string        `plist:"_parentDataType"`
+
+	// These would be nice to add, but they come back with inconsistent
+	// types, so doing a straight unmarshal is hard.
+	// DetailLevel    int                 `plist:"_detailLevel"`
+	// Properties     map[string]Property `plist:"_properties"`
+}
+
+type Table struct {
+	client    *osquery.ExtensionManagerClient
+	logger    log.Logger
+	tableName string
+}
+
+func TablePlugin(client *osquery.ExtensionManagerClient, logger log.Logger) *table.Plugin {
+
+	columns := []table.ColumnDefinition{
+		table.TextColumn("fullkey"),
+		table.TextColumn("parent"),
+		table.TextColumn("key"),
+		table.TextColumn("value"),
+		table.TextColumn("parentdatatype"),
+
+		table.TextColumn("query"),
+		table.TextColumn("datatype"),
+		table.TextColumn("detaillevel"),
+	}
+
+	t := &Table{
+		client:    client,
+		logger:    level.NewFilter(logger, level.AllowInfo()),
+		tableName: "kolide_system_profiler",
+	}
+
+	return table.NewPlugin(t.tableName, columns, t.generate)
+}
+
+func (t *Table) generate(ctx context.Context, queryContext table.QueryContext) ([]map[string]string, error) {
+	var results []map[string]string
+
+	requestedDatatypes := []string{}
+
+	datatypeQ, ok := queryContext.Constraints["datatype"]
+	if !ok || len(datatypeQ.Constraints) == 0 {
+		return results, errors.Errorf("The %s table requires that you specify a constraint for datatype", t.tableName)
+	}
+
+	for _, datatypeConstraint := range datatypeQ.Constraints {
+		dt := datatypeConstraint.Expression
+
+		// If the constraint is the magic "%", it's eqivlent to an `all` style
+		if dt == "%" {
+			requestedDatatypes = []string{}
+			break
+		}
+
+		requestedDatatypes = append(requestedDatatypes, dt)
+	}
+
+	var detailLevel string
+	if q, ok := queryContext.Constraints["detaillevel"]; ok && len(q.Constraints) != 0 {
+		if len(q.Constraints) > 1 {
+			level.Info(t.logger).Log("msg", "WARNING: Only using the first detaillevel request")
+		}
+
+		dl := q.Constraints[0].Expression
+		for _, known := range knownDetailLevels {
+			if known == dl {
+				detailLevel = dl
+			}
+		}
+
+	}
+
+	systemProfilerOutput, err := t.execSystemProfiler(ctx, detailLevel, requestedDatatypes)
+	if err != nil {
+		return results, errors.Wrap(err, "exec")
+	}
+
+	if q, ok := queryContext.Constraints["query"]; ok && len(q.Constraints) != 0 {
```

Comment:
Wouldn't it be simpler to avoid trying to parse the rows out of the output in the table generate func by using `system_profiler -json -detailLevel <level> [dataType1 ... dataTypeN]` just returning the JSON back. The consumer can then use sqlite's JSON1 extension to query the output or not.

Reply:
I believe @directionless mentioned to me that the `-json` flag is not supported/available on versions of macOS prior to 10.15 _(Catalina)_

Reply:
I'm not totally sure I understand your question.

As fritz mentions, `-json` is a catalina option. But I could work around that. I think the underlying question here is what does datatable's `query` do that sqlite's `json1` doesn't already do? If that's it, I think that's a great question. 

I wrote dataflatten to be able to take arrays of maps, and re-write them into a map. Thus allowing simpler manipulations in EAV style tables. When I wrote it, I couldn't find good prior art that felt right. xpath is the right idea. but big and hard to use in a simple select interface. Similar for some existing json things.

I did not evaluate sqlite's `json1` one at the time, entirely because I wasn't aware of it :(

Looking at it now, I'm not sure it works. While it handles some use cases, we have a pretty common case of taking the aforementioned array-of-maps, converting it into a nested map, and then pivoting it into a wide tuple. `json_extract` might do that, but I don't immediately see how. 

I'm happy to consider the json1 stuff in the future. Though I think there several reasons to kick that forward:
* unsure about the feasibility, we'd need to play with some of our queries
* We'd need to make a general plist -> json routine
* We already have this